include/crypto/lms.h

```diff
@@ -0,0 +1,146 @@
+#  define OSSL_LMS_TYPE_SHAKE_N24_H20   0x00000017
+#  define OSSL_LMS_TYPE_SHAKE_N24_H25   0x00000018
+
+#  define OSSL_LM_OTS_TYPE_SHA256_N32_W1 0x00000001
+#  define OSSL_LM_OTS_TYPE_SHA256_N32_W2 0x00000002
+#  define OSSL_LM_OTS_TYPE_SHA256_N32_W4 0x00000003
+#  define OSSL_LM_OTS_TYPE_SHA256_N32_W8 0x00000004
+#  define OSSL_LM_OTS_TYPE_SHA256_N24_W1 0x00000005
+#  define OSSL_LM_OTS_TYPE_SHA256_N24_W2 0x00000006
+#  define OSSL_LM_OTS_TYPE_SHA256_N24_W4 0x00000007
+#  define OSSL_LM_OTS_TYPE_SHA256_N24_W8 0x00000008
+#  define OSSL_LM_OTS_TYPE_SHAKE_N32_W1  0x00000009
+#  define OSSL_LM_OTS_TYPE_SHAKE_N32_W2  0x0000000A
+#  define OSSL_LM_OTS_TYPE_SHAKE_N32_W4  0x0000000B
+#  define OSSL_LM_OTS_TYPE_SHAKE_N32_W8  0x0000000C
+#  define OSSL_LM_OTS_TYPE_SHAKE_N24_W1  0x0000000D
+#  define OSSL_LM_OTS_TYPE_SHAKE_N24_W2  0x0000000E
+#  define OSSL_LM_OTS_TYPE_SHAKE_N24_W4  0x0000000F
+#  define OSSL_LM_OTS_TYPE_SHAKE_N24_W8  0x00000010
+
+/* XDR sizes when encoding and decoding */
+#  define LMS_SIZE_I 16
+#  define LMS_SIZE_LMS_TYPE 4
+#  define LMS_SIZE_OTS_TYPE 4
+
+/*
+ * Refer to RFC 8554 Section 4.1.
+ * See also lm_ots_params[]
+ */
+typedef struct lm_ots_params_st {
+    /*
+     * The OTS type associates an id with a set of OTS parameters
+     * e.g. OSSL_LM_OTS_TYPE_SHAKE_N32_W1
+     */
+    uint32_t lm_ots_type;
+    uint32_t n;              /* Hash output size in bytes (32 or 24) */
+    /*
+     * The width of the Winternitz coefficients in bits. One of (1, 2, 4, 8)
+     * Higher values of w are slower (~2^w computations) but have smaller
+     * signatures.
+     */
+    uint32_t w;
+    /*
+     * The number of n-byte elements used for an LMOTS signature.
+     * One of (265, 133, 67, 34) for n = 32, for w=1,2,4,8
+     * One of (200, 101, 51, 26) for n = 24, for w=1,2,4,8
+     */
+    uint32_t p;
+    const char *digestname; /* Hash Name */
+} LM_OTS_PARAMS;
+
+/* See lms_params[] */
+typedef struct lms_params_st {
+    /*
+     * The lms type associates an id with a set of parameters to define the
+     * Digest and Height of a LMS tree.
+     * e.g, OSSL_LMS_TYPE_SHA256_N24_H25
+     */
+    uint32_t lms_type;
+    const char *digestname; /* One of SHA256, SHA256-192, or SHAKE256 */
+    uint32_t n; /* The Digest size (either 24 or 32), Useful for setting up SHAKE */
+    uint32_t h; /* The height of a LMS tree which is one of 5, 10, 15, 20, 25) */
+} LMS_PARAMS;
+
+typedef struct lms_pub_key_st {
+    /*
+     * A buffer containing an encoded public key of the form
+     * u32str(lmstype) || u32str(otstype) || I[16] || K[n]
+     */
+    unsigned char *encoded;         /* encoded public key data */
+    size_t encodedlen;
+    /*
+     * K is the LMS tree's root public key (Called T(1))
+     * It is n bytes long (the hash size).
+     * It is a pointer into the encoded buffer
+     */
+    unsigned char *K;
+    uint32_t allocated;             /* If 1 then encoded needs to be freed */
```

Comment:
What are the cases when it is not allocated?

Reply:
There doesn't appear to be any.  @slontis, okay to remove this or does it get used as part of HSS?